### PR TITLE
[WIP] ScanPages/ScanPagesWithContext support.

### PR DESCRIPTION
**Pull request by zabawaba99**
*Issue #, if available:*
* fixes aws/aws-dax-go#3

*Description of changes:*
* Add support for ScanPages/ScanPagesWithContext

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: dax/api.go [hunk 1]
```diff
@@ -16,11 +16,12 @@
 package dax
 
 import (
+	"io"
```
Comment (zabawaba99) [inline dax/api.go]: I'll revert this as this change has nothing to do with the fix - it was something that my formatter did.

File: dax/api.go [hunk 2]
```diff
@@ -247,12 +248,27 @@ func (d *Dax) QueryPagesWithContext(aws.Context, *dynamodb.QueryInput, func(*dyn
 	return d.unImpl()
 }
 
-func (d *Dax) ScanPages(*dynamodb.ScanInput, func(*dynamodb.ScanOutput, bool) bool) error {
-	return d.unImpl()
+func (d *Dax) ScanPages(in *dynamodb.ScanInput, fn func(*dynamodb.ScanOutput, bool) bool) error {
```
Comment (zabawaba99) [inline dax/api.go]: For consistency, `in` should be renamed to `input`.

File: dax/api.go [hunk 3]
```diff
@@ -247,12 +248,27 @@ func (d *Dax) QueryPagesWithContext(aws.Context, *dynamodb.QueryInput, func(*dyn
...
+	return d.ScanPagesWithContext(nil, in, fn)
 }
 
-func (d *Dax) ScanPagesWithContext(aws.Context, *dynamodb.ScanInput, func(*dynamodb.ScanOutput, bool) bool, ...request.Option) error {
-	return d.unImpl()
+func (d *Dax) ScanPagesWithContext(ctx aws.Context, in *dynamodb.ScanInput, fn func(*dynamodb.ScanOutput, bool) bool, opts ...request.Option) error {
```
Comment (zabawaba99) [inline dax/api.go]: For consistency, `in` should be renamed to `input`.

File: dax/api.go [hunk 4]
```diff
@@ -247,12 +248,27 @@ func (d *Dax) QueryPagesWithContext(aws.Context, *dynamodb.QueryInput, func(*dyn
...
+	doWork = func() error {
+		out, err := d.ScanWithContext(ctx, in, opts...)
+		if err != nil {
+			return err
+		}
+
+		isLastPage := out.LastEvaluatedKey != nil
+		cont := fn(out, isLastPage)
+		if cont && !isLastPage {
+			in.ExclusiveStartKey = out.LastEvaluatedKey
```
Comment (zabawaba99) [inline dax/api.go]: We should probably be making a copy of this on L#256 so that we don't modify the user input.

File: dax/api.go [hunk 5]
```diff
@@ -247,12 +248,27 @@ func (d *Dax) QueryPagesWithContext(aws.Context, *dynamodb.QueryInput, func(*dyn
...
-	return d.unImpl()
+func (d *Dax) ScanPagesWithContext(ctx aws.Context, in *dynamodb.ScanInput, fn func(*dynamodb.ScanOutput, bool) bool, opts ...request.Option) error {
+	var doWork func() error
+	doWork = func() error {
+		out, err := d.ScanWithContext(ctx, in, opts...)
+		if err != nil {
+			return err
+		}
+
+		isLastPage := out.LastEvaluatedKey != nil
```
Comment (jiaqihu-A) [inline dax/api.go]: Should this be "LastEvaluatedKey == nil"?

**Comment by aidansteele** [comment]
Is there an ETA on this getting merged?

**Comment by zabawaba99** [comment]
Hey @aidansteele @jiaqihu-A I apologize for not getting to the comments earlier; let me get these addressed.

**Comment by zabawaba99** [comment]
I just saw that @aidansteele submitted #15. Might as well go with that one as it handles queries as well.

**Pull request closed without merging**